lib/extensions.js

```diff
@@ -40,6 +40,34 @@ module.exports = function(realmConstructor) {
     setConstructorOnPrototype(realmConstructor.Results);
     setConstructorOnPrototype(realmConstructor.Object);
 
+    //Add async open API
+    Object.defineProperties(realmConstructor, getOwnPropertyDescriptors({
+        open(config) {
+            return new Promise((resolve, reject) => {
+                const realm = new realmConstructor(config);
```

Comment:
I don't think that's correct. This will open a new Realm instance with the schema which will push to the server. Cocoa actually creates a special Realm that doesn't own a schema and wait on its sync session before creating a regular Realm - https://github.com/realm/realm-cocoa/blob/master/Realm/RLMRealm.mm#L193.

I believe what we ought to do is add a `Realm.Sync._getSession(syncConfig)` which maps to `SyncManager::get_session)`, call `_waitForDownload(callback)` on the Session and in the callback resolve the promise with a new Realm. Then this code becomes:
```js
const session = realmConstructor.Sync._getSession(config.sync);
session._waitForDownload((error) => {
    if (error) reject(error);
    else resolve(new realmConstructor(config));
});
```